### ISSUE TITLE
supabase start errors since migration, can't run supabase locally any more

### ISSUE THREAD

**Issue by barbalex** [supabase/supabase#7425]
# Bug report

## Describe the bug

```
PS C:\Users\alexa\capturing> supabase start
Error: Error starting database: ERROR:  relation "project_users" does not exist
LINE 15:         INNER JOIN project_users ON users.email = project_us...
                            ^
ERROR:  current transaction is aborted, commands ignored until end of transaction block
```

I uninstalled supabase, reinstalled it. Did not help.

What can I do to be able to work on my project again?

## System information

- OS: Windows 11 up to date
- Version of supabase-js: newest, just installed
- Version of Node.js: 17.3.1

**Comment by barbalex** [comment]
I _think_ I understand what is happening: Problems started after generating a migration. I did:

```
supabase db commit initiated
```

I now found out that there is a `supabase` folder in my project, containing a `migrations` folder. and the file `20220625100132_initiated.sql` contained in there has this as first lines:
```sql

CREATE OR REPLACE FUNCTION public.is_project_user_by_vector_layer(
	_auth_user_id uuid,
	_vector_layer_id uuid)
    RETURNS boolean
    LANGUAGE 'sql'
    COST 100
    VOLATILE SECURITY DEFINER PARALLEL UNSAFE
AS $BODY$
  SELECT
    EXISTS (
      SELECT
        1
      FROM
        auth.users users
        INNER JOIN project_users ON users.email = project_users.user_email
        INNER JOIN projects ON projects.id = project_users.project_id
        INNER JOIN vector_layers ON vector_layers.project_id = projects.id
      WHERE
        vector_layers.id = _vector_layer_id
        AND users.id = _auth_user_id);

$BODY$;
```

This of course will not work, as a function depending on tables is built _before_ the tables are built.

So I guess that explains the error?

Problem is: My own script to generate a new db is about 2700 lines long. And I worked very long to get the sequence of events right so no errors would occur because things are used before being defined. That is the script I run before trying to create the migration.

`20220625100132_initiated.sql` contains **3644** lines of sql.

Am I expected to go through this code and check every line to find too early usage of stuff? 😱 

What when during the development of my project struckture is changed, which happens quite often? My custom is to keep my own sql clean. But rebuilding migrations would then mean that I have to _again_ refactor 3644 lines of sql!!!

Plus: I often get local dev errors because the docker instance runs out of disk space. This is VERY annoying and means that supabase has to be stopped and restarted. Without migrations that means a complete rebuild of everything.

So how can I get my project to work when using supabase?

I must be missing something basic.

**Comment by barbalex** [comment]
Hm. The name `Schema Diff utility` implies that my carefully created sql was changed to find differences to a previous version.

It seem to me this does not make sense when a project is initialized. There is nothing to compare. So better use the code input by the dev.

Am I missing something?

**Comment by barbalex** [comment]
Idea: can I simply replace `20220625100132_initiated.sql` with my own script, https://github.com/barbalex/capturing/blob/main/src/sql/tables.sql? Maybe adding the shema name to table and function names?

**Comment by kiwicopple** [comment]
> can I simply replace 20220625100132_initiated.sql with my own script

Yes, you could - the generated schema is just a diff (and a "best guess"), which you're recommended to edit to ensure it fits your needs before you check it in to VC. You can create and edit migration file as you see fit. In fact, you don't _have_ to use supabase CLI for migrations - you can use the tool which you've been using previously. Supabase is just PostgreSQL, so everything that works with postgres also works with supabase

**Comment by barbalex** [comment]
I am getting a bit frustrated, replacing the migration with my own. One of the problems is that I got literally hundreds of errors claiming tables did not exist. As I now control the sequence of events I thought that one difference in the migration could be that the migration required qualified names i.e. `public.rows` instead of `rows`.

Hours later, after converting many names, I now am confronted with a very long list of claims, that qualified names did not exist.

Either the fact that this is such a long script gets me. Or the series of events is changed. To be honest, I do not know how to go on.

**Comment by barbalex** [comment]
This is the errors listed when I use my original sql file tested many times as migration file:
```
PS C:\Users\alexa\capturing> supabase start
Error: Error starting database: ERROR:  type "geometry" does not exist
LINE 5:   geometry geometry(GeometryCollection, 4326) DEFAULT NULL,
                   ^
ERROR:  relation "rows" does not exist
ERROR:  relation "rows" does not exist
ERROR:  relation "rows" does not exist
ERROR:  relation "rows" does not exist
ERROR:  relation "rows" does not exist
ERROR:  relation "rows" does not exist
ERROR:  relation "rows" does not exist
ERROR:  relation "rows" does not exist
ERROR:  relation "rows" does not exist
ERROR:  relation "rows" does not exist
ERROR:  relation "rows" does not exist
ERROR:  relation "rows" does not exist
ERROR:  relation "rows" does not exist
ERROR:  relation "rows" does not exist
ERROR:  relation "rows" does not exist
ERROR:  relation "rows" does not exist
ERROR:  relation "rows" does not exist
ERROR:  relation "rows" does not exist
ERROR:  relation "rows" does not exist
ERROR:  type "geometry" does not exist
LINE 6:   geometry geometry(GeometryCollection, 4326) DEFAULT NULL,
                   ^
ERROR:  relation "row_revs" does not exist
ERROR:  relation "row_revs" does not exist
ERROR:  relation "row_revs" does not exist
ERROR:  relation "row_revs" does not exist
ERROR:  relation "row_revs" does not exist
ERROR:  relation "row_revs" does not exist
ERROR:  relation "row_revs" does not exist
ERROR:  relation "row_revs" does not exist
ERROR:  relation "row_revs" does not exist
ERROR:  relation "row_revs" does not exist
ERROR:  relation "row_revs" does not exist
ERROR:  relation "row_revs" does not exist
ERROR:  relation "row_revs" does not exist
ERROR:  relation "row_revs" does not exist
ERROR:  relation "row_revs" does not exist
ERROR:  relation "row_revs" does not exist
ERROR:  relation "row_revs" does not exist
ERROR:  relation "rows" does not exist
ERROR:  relation "files_meta" does not exist
ERROR:  relation "files_meta" does not exist
ERROR:  relation "files_meta" does not exist
ERROR:  relation "files_meta" does not exist
ERROR:  relation "files_meta" does not exist
ERROR:  relation "rows" does not exist
ERROR:  relation "files_meta" does not exist
ERROR:  relation "files_meta" does not exist
ERROR:  relation "files_meta" does not exist
ERROR:  relation "files_meta" does not exist
ERROR:  relation "files_meta" does not exist
ERROR:  relation "files_meta" does not exist
ERROR:  relation "files_meta" does not exist
ERROR:  relation "files_meta" does not exist
ERROR:  relation "files_meta" does not exist
ERROR:  relation "files_meta" does not exist
ERROR:  relation "files_meta" does not exist
ERROR:  relation "files_meta" does not exist
ERROR:  type "geometry" does not exist
LINE 4:   geometry geometry(GeometryCollection, 4326) DEFAULT NULL,
                   ^
ERROR:  relation "pvl_geoms" does not exist
ERROR:  relation "pvl_geoms" does not exist
ERROR:  relation "pvl_geoms" does not exist
ERROR:  relation "pvl_geoms" does not exist
ERROR:  relation "pvl_geoms" does not exist
ERROR:  relation "pvl_geoms" does not exist
ERROR:  relation "pvl_geoms" does not exist
ERROR:  relation "pvl_geoms" does not exist
ERROR:  relation "pvl_geoms" does not exist
ERROR:  relation "pvl_geoms" does not exist
ERROR:  relation "pvl_geoms" does not exist
ERROR:  relation "pvl_geoms" does not exist
ERROR:  relation "rows" does not exist
LINE 13:         INNER JOIN ROWS ON rows.table_id = tables.id
                            ^
ERROR:  relation "rows" does not exist
LINE 13:         INNER JOIN ROWS ON rows.table_id = tables.id
                            ^
ERROR:  relation "rows" does not exist
LINE 13:         INNER JOIN ROWS ON rows.table_id = tables.id
                            ^
ERROR:  relation "rows" does not exist
LINE 13:         INNER JOIN ROWS ON rows.table_id = tables.id
                            ^
ERROR:  relation "rows" does not exist
ERROR:  relation "rows" does not exist
ERROR:  relation "rows" does not exist
ERROR:  relation "rows" does not exist
ERROR:  relation "row_revs" does not exist
ERROR:  relation "row_revs" does not exist
ERROR:  relation "row_revs" does not exist
ERROR:  relation "row_revs" does not exist
ERROR:  relation "files_meta" does not exist
ERROR:  relation "files_meta" does not exist
ERROR:  relation "files_meta" does not exist
ERROR:  relation "files_meta" does not exist
ERROR:  relation "pvl_geoms" does not exist
ERROR:  relation "pvl_geoms" does not exist
ERROR:  relation "pvl_geoms" does not exist
ERROR:  relation "pvl_geoms" does not exist
```

**Comment by barbalex** [comment]
I just realized: `type "geometry" does not exist` now also happens now when I run my original sql file in the local studio or via pgadmin.

So it seems that somehow the installation of supabase went wrong?

Will uninstall and re-install again.

**Comment by barbalex** [comment]
I did `scoop uninstall supabase`, then `soop install supabase`.  Postgis has not been installed, again.

**Comment by barbalex** [comment]
Because postis is not present there is no way for me to work with supabase 😢

**Comment by kiwicopple** [comment]
> there is no way for me to work with supabase any more 😢

Sorry to hear that - I completely understand though as it looks like there has been a lot of effort spent here.

I'd love to nail this for the future though. I noticed that the postgis installation is commented out here:

https://github.com/barbalex/capturing/blob/f40259a50558fb4594b3c45f78e72be6eef08760/src/sql/tables.sql#L5

Is that intentional?

To replicate your issue would I do this? --

1. `supabase init`
2. copy your [tables.sql](https://github.com/barbalex/capturing/blob/f40259a50558fb4594b3c45f78e72be6eef08760/src/sql/tables.sql) script into the supabase init migration
3. run `supabase start` 

Let me know if those are the steps you're taking and I'll make sure we fix this

**Comment by barbalex** [comment]
I learned some things that seem to have been the reason for these issues:
- Because I have my own `users` table my sql code contained: `DROP TABLE IF EXISTS users CASCADE;`. As this ran one line before the users table was created, the table that was going to be deleted _if my own users table had never been created before_ was `auth.users`. I corrected it by adding the schema name: `DROP TABLE IF EXISTS public.users CASCADE;`
- While debugging I outcommented the postgis call `CREATE EXTENSION IF NOT EXISTS postgis;`. And forgot to reactivate it
- Because I had had other issues earlier, there were a few unnecessary calls that I do not use often and the implications of which I did not completely understand, like:
  - `start transaction`, later `stop transaction`
  - `SET CONSTRAINTS ALL DEFERRED;`
  - `BEGIN` and `COMMIT` around `CREATE publication supabase_realtime;`

Now that this has all been corrected, the manually built migration works like a charm.

Sorry for causing so much trouble.

I appreciate your help VERY much. Supabase is a great team with a great product. Congratulations!
